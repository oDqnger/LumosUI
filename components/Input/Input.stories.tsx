import { Meta, StoryObj } from "@storybook/react";
import Input from "./Input"
import "./../../styles/output.css"

const meta: Meta<typeof Input> = {
    component: Input,
}

export default meta;
type Story = StoryObj<typeof Input>;

export const DefaultInput: Story = {
    args: {
        disabled: true,
        type: "text",
        readOnly: false,
        required: false,
        sizes:"md",
        radius:"md"
    }
}
