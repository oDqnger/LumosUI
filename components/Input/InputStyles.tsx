export const sizeStyles = {
    "sm": " pt-2 pb-2 pl-5 pr-5 text-sm",
    "md": " pt-3 pb-3 pl-7 pr-7 text-base",
    "lg": " pt-4 pb-4 pl-9 pr-9 text-lg"
}

export const radiusStyles = {
    "full": " rounded-full",
    "lg": " rounded-lg",
    "md": " rounded-md",
    "sm": " rounded-sm",
    "none": " rounded-none"
}

export const baseStyle = " border-2 font-semibold text-white transition ease-in-out hover:bg-gray-600 border-gray-500 bg-gray-500"
export const disabledStyle = " border-gray-800 bg-gray-800";
export const readOnlyStyle = " transition ease-in-out bg-gray-500 border-gray-600 hover:border-gray-700"
