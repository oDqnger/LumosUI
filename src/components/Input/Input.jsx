export default function Input(props) {

    const {
        type,
        label,
        placeholder,
        autoComplete,
        autocapitalize,
        dirname,
        disabled,
        form,
        maxLength,
        minLength,
        max,
        min,
        multiple,
        name,
        pattern,
        readOnly,
        required,
        size,
        defaultValue,
        sizes,
        radius,
        className,
        ...defaultProps
    } = props;

    const inputTypes = [
        "email",
        "number",
        "password",
        "search",
        "tel",
        "text",
        "url",
    ]

    const sizeStyles = {
        "sm": "pt-2 pb-2 pl-5 pr-5 text-sm",
        "md": "pt-3 pb-3 pl-7 pr-7 text-base",
        "lg": "pt-4 pb-4 pl-9 pr-9 text-lg"
    }

    const [radiusStyles, selectRadius] = [{
        "full": "rounded-full",
        "lg": "rounded-lg",
        "md": "rounded-md",
        "sm": "rounded-sm",
        "none": "rounded-none"
    }, () => radiusStyles[radius] ?? radiusStyles["lg"]]

    const baseStyle = className + `border-2 ${selectRadius()} ${sizeStyles[sizes] ?? sizeStyles["md"]} font-semibold text-white`;
    const normalStyle = baseStyle + " transition ease-in-out hover:bg-gray-600 border-gray-500 bg-gray-500";
    const disabledStyle = baseStyle + " border-gray-800 bg-gray-800";
    const readOnlyStyle = baseStyle + " transition ease-in-out bg-gray-500 border-gray-600 hover:border-gray-700"

    const selectedStyle = () => {
        if (disabled) {
            return disabledStyle;
        } else if (readOnly) {
            return readOnlyStyle;
        } else {
            return normalStyle;
        }
    }

    return (
        <>
            <input
            className={selectedStyle()}
            type={inputTypes.includes(type) ? type : "text"}
            placeholder={placeholder}
            aria-label={label}
            defaultValue={defaultValue}
            disabled={disabled}
            readOnly={readOnly}
            required={required}
            />
        </>
    )
}
