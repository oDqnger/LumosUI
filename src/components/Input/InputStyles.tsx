/* eslint-disable @typescript-eslint/no-unused-vars */

export const sizeStyles = {
    "sm": "pt-2 pb-2 pl-5 pr-5 text-sm",
    "md": "pt-3 pb-3 pl-7 pr-7 text-base",
    "lg": "pt-4 pb-4 pl-9 pr-9 text-lg"
}

export const radiusStyles = {
    "full": "rounded-full",
    "lg": "rounded-lg",
    "md": "rounded-md",
    "sm": "rounded-sm",
    "none": "rounded-none"
}
