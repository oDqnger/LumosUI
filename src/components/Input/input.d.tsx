import { HTMLAttributes } from "react";

type InputTypes = "email" | "number" | "password" | "search" | "tel" | "text" | "url";
type SizeTypes = "sm" | "md" | "lg"
type RadiusTypes = "none" | "sm" | "md" | "lg" | "full"

declare interface InputProps extends HTMLAttributes<HTMLInputElement> {
    type: InputTypes,
    disabled?: boolean,
    readOnly?: boolean,
    required?: boolean,
    className?: string,
}

declare interface StyledProps {
    sizes?: SizeTypes,
    radius?: RadiusTypes,
}

export type AllProps = InputProps & StyledProps;
