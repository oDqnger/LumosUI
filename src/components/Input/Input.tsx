import React, { forwardRef } from "react";
import { sizeStyles, radiusStyles, baseStyle, readOnlyStyle, disabledStyle } from "./InputStyles";
import { AllProps } from "./input.d"

function Input(props: AllProps, ref) {

    const {
        type,
        label,
        placeholder,
        autoComplete,
        autocapitalize,
        dirname,
        disabled,
        form,
        maxLength,
        minLength,
        max,
        min,
        multiple,
        name,
        pattern,
        readOnly,
        required,
        defaultValue,
        sizes="md",
        radius="md",
        className,
        ...defaultProps
    } = props;

    const selectDisabled = () => disabled ? disabledStyle : "";
    const selectReadOnly = () => readOnly ? readOnlyStyle : "";

    const combinedStyles = 
    className
    +baseStyle
    +radiusStyles[radius]
    +sizeStyles[sizes]
    +selectDisabled()
    +selectReadOnly()

    return (
        <>
            <input
            className={combinedStyles}
            type={type}
            placeholder={placeholder}
            aria-label={label}
            defaultValue={defaultValue}
            disabled={disabled}
            readOnly={readOnly}
            required={required}
            ref={ref}
            { ...defaultProps }
            />
        </>
    )
}

export default forwardRef(Input);
