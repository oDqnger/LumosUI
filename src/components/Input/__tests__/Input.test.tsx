import '@testing-library/jest-dom/extend-expect';
import React from "react"
import { render } from "@testing-library/react"
import Input from "../Input"

describe('Input Component', () => {
    it("should render input component", () => {
        const { getByLabelText } = render(<Input label="test" type="email" />)
        // eslint-disable-next-line testing-library/prefer-screen-queries
        const inputBtn = getByLabelText("test");
        expect(inputBtn).toBeInTheDocument();
    })

    // it("should use input types accordingly", () => {
    //     const inputTypes = ["email", "number", "password", "search", "tel", "text", "url"];

    //     for (let typess of inputTypes) {
    //         let { getByLabelText } = render(<Input label="test" type={typess}></Input>)
    //         expect(getByLabelText("test")).toBeInTheDocument();
    //     }
    // })
})