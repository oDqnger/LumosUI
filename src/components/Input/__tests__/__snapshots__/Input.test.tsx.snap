// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match the snapshot 1`] = `
<div>
  <input
    class="undefined border-2 rounded-md pt-3 pb-3 pl-7 pr-7 text-base font-semibold text-white transition ease-in-out hover:bg-gray-600 border-gray-500 bg-gray-500"
    type="email"
    value=""
  />
</div>
`;
